### ISSUE TITLE
Cluster Datastore unmarshalling bug on shrink operation

### ISSUE THREAD

**Issue by gozer** [containous/traefik#2153]
### Do you want to request a *feature* or report a *bug*?

This is a bug in cluster/datastore that causes ACME's state to be unreadable/corrupted during cluster operations

### What did you do?

Running Traefik in clustered mode with Consul as the ACME storage backend.

Ocasionally, we would see erratic behaviour that ended up with ACME registration/certificate issuance failing and being retried for no apparent good reason.

The issue ultimately bubbled up once we reached let's encrypt rate limts and were effetively dead in the water

### What did you expect to see?

Stable state and no unnecessary ACME certificate reissuance attempts

### What did you see instead?

Apart from the error logged below, we were effectively seeing very slow TLS initial negociation intermittently and sometimes TLS negociation failures. All consistent with Let's Encrypt re-issuance

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.3.7
Codename:     raclette
Go version:   go1.8.3
Built:        2017-08-25_08:36:23PM
OS/Arch:      linux/amd64

```

```toml
# (paste your configuration here)
```

### If applicable, please paste the log output in debug mode (`--debug` switch)

```
ERRO[2017-09-20T17:21:41Z] Error reloading: invalid character 'c' after top-level value 
ERRO[2017-09-20T17:21:41Z] Error in watch datastore: invalid character 'c' after top-level value, retrying in 579.298905ms
```

**Comment by gozer** [comment]
Tracked down the issue to a very interesting case of unitialized memory, and my apologies if my go-foo isn't very good.

Reproducing the issue was a bit tricky but can actually be done relatively easily. I've been using the following approach, but I am sure there can be more succint ones.

Basically, if the ACME's stored config object in Consul *shrinks* in size, what comes back out from cluster/datastore.go still contains the trailing portion of the config object from *before*, i.e.

```
{ email: 'gozer@ectoplasm.org' }
 .. (after setting it to longer)
{ email: 'gozer-longer@ectoplasm.org' }
... (after setting it back to short)
{ email: 'gozer-longer@ectoplasm.org'}m.org' }
```

My poor-man's reproduction test case was:

```shell
consulate kv set acme/object "$(cat big | base64)" ; 
consulate kv set acme/lock $RANDOM; 
sleep 1;  
consulate kv set acme/object "$(cat small | base64)"; 
consulate kv set acme/lock $RANDOM
```
And I could trigger the bug every time.

**Comment by gozer** [comment]
I'd like to make a PR to fix this, having managed to track it down to staert's use of mitchellh/mapstructure without the `ZeroFields:       true` option.

However, I am not entirely certain I fully understand the ramifications of the possible fixes I've found, and I'd like some guidance/opinions.

Internally to traefik, we can fix it with

```diff
diff --git a/cluster/datastore.go b/cluster/datastore.go
index 538d70be..fe4cc617 100644
--- a/cluster/datastore.go
+++ b/cluster/datastore.go
@@ -121,13 +129,24 @@ func (d *Datastore) watchChanges() error {
 func (d *Datastore) reload() error {
        log.Debugf("Datastore reload")
        d.localLock.Lock()
+
+       // clear Object first, as mapstructure's decoder doesn't have ZeroFields set to true
+       d.meta.Object = d.meta.Object[:0]
+
+       //log.Debugf("Meta before %s", d.meta.Object)
+
        err := d.kv.LoadConfig(d.meta)

        if err != nil {
                d.localLock.Unlock()
                return err
        }
        err = d.meta.unmarshall()
        if err != nil {
                d.localLock.Unlock()
                return err
        }

```

Or what feels like a better fix to me, but lives in containous/staert would be like 

```diff
diff --git a/vendor/github.com/containous/staert/kv.go b/vendor/github.com/containous/staert/kv.go
index 528f0a59..7d2a9841 100644
--- a/vendor/github.com/containous/staert/kv.go
+++ b/vendor/github.com/containous/staert/kv.go
@@ -52,6 +52,7 @@ func (kv *KvSource) LoadConfig(config interface{}) error {
        }
        // fmt.Printf("mapstruct : %#v\n", mapstruct)
        configDecoder := &mapstructure.DecoderConfig{
+               ZeroFields:       true,
                Metadata:         nil,
                Result:           config,
                WeaklyTypedInput: true,

```

**Comment by gozer** [comment]
But to summarize, cluster/datastore persists data to Consul on user's behalf, but when that data *shrinks* for whatever reason, it might fail to decode it back.

In our case, the culprit was failed ACME validation requests, that caused the ACME config to grow with the new cert then shrink back down once the validation request ultimately failed.

**Comment by emilevauge** [comment]
:clap: :clap: :clap: for this awesome and deep analyze !

**Comment by nmengin** [comment]
@gozer Many thanks for this complete analyze!!!

The best thing to do is certainly opening a new PR on [stært](https://github.com/containous/staert/pulls) in the way to use the modified version in the next Træfik release.

**Comment by gozer** [comment]
@nmengin I've thought about this some more and I've opened the initial issue against staert, but it got me thinking, and maybe someone that knows how staert is used could shed some light.

I believe my suggested patch to staert  `ZeroFields: true` would have unintended consequences to staert's configuration merging between different backing stores, no?

I believe not having ZeroFields set is what allows staert to possibly merge bits of configuration from different sources into the same configuration object.

For instance, if I had in TOML `{ foo: { a: 123} }` and in the KV `{ foo: { b: 456} }`, wouldn't the expected final state for the object be `{ foo: { a: 123, b: 456 }}` ?

With that patch, I believe only the last writer would win.

If that's the case, then, the correct solution would be to explicitely avoid that merging for just that one specific case of ACME's storage of a JSON string representation as an array of bytes[]

```
d.meta.Object = d.meta.Object[:0]
```

Or to be possibly cleaner and do something like the following pseudo-code:

```
func (d *Datastore) reload() error {
        d.localLock.Lock()

        newmeta := make(Metadata)

        err := d.kv.LoadConfig(newmeta)

        if err != nil {
                d.localLock.Unlock()
                return err
        }
        err = newmeta.unmarshall()

    if err != nil {
                d.localLock.Unlock()
                return err
        }
   // Swap in our new, valid, config object
    d.meta = newmeta

```

**Comment by traefiker** [comment]
Closed by #2165.
